### ISSUE TITLE
Possibility to read password from keyring, gpg encrypted file or script

### ISSUE THREAD

**Issue by x4121** [akrennmair/newsbeuter#502]
Tools like mutt, OfflineIMAP, mbsync, msmtp allow you to read the password from keyrings or gpg encrypted files (or some arbitrary script).
If I didn't miss anything, Newsbeuter only reads passwords from plain text files. Is there a similar feature planed for Newsbeuter or is this not wanted or not possible with the current codebase?

**Comment by Minoru** [comment]
I wasn't even aware such features exist in other programs! Not in the ones aimed at text consoles, at least.

I have nothing against this, and I don't think there's anything in our codebase that prevents us from implementing this. You're welcome to look into this if you have an inclination.

**Comment by x4121** [comment]
I guess the easiest approach would be the one of msmtp. They have a config parameter `password` that expects a plain text password and `passwordeval` that expects an external command to return the password (something like `passwordeval "secret-tool lookup email my-email"`). This wouldn't limit the use to gpg, the system keyring or python scripts (as does OfflineIMAP). msmtp is written in C, so maybe this could be used as "inspiration".

**Comment by x4121** [comment]
I did a simple implementation for `passwordeval`, I put it (with the "read password from file") in a separate class, so it can be reused in other places.
Will also add this for the oder parts containing passwords, write some documentation and create a PR tomorrow.

I wouldn't consider myself a "C Developer", so any input/critique is appreciated ;)

**Comment by Minoru** [comment]
Wow, you're fast! I'd rather you submit the PR right away and we review the code there—that way it'll be easier to find the code and the discussion later on, and it's guaranteed to be preserved as long as *this* repo exists.

**Comment by x4121** [comment]
OK, created the PR as #503

**Comment by Minoru** [comment]
Doing a little research here to better understand how other programs do this. Read about OfflineIMAP's `remotepasswordeval` and msmtp's `passwordeval`, but can't find anything of the sort for Mutt. What's the setting called there?

The thing is, I don't like the name "eval" and would prefer "cmd" or "command", but OfflineIMAP and msmtp use "eval", so maybe we should go with it to to make it easier to find.

**Comment by x4121** [comment]
In Mutt you just have to put the statement in backticks instead of quotes. But IMHO this is not as straight forward, this might be familiar to programmers but could be confusing to some less technical users.

I personally liked the term "eval" since this emphasizes on the evaluation/result of the given statement, not on the execution of some command.
But this is just a personal preference, I'm open for debate here ;)

Had a look at other tools I use, just recognized that not a lot of them require authentication or do this via own keyrings/agents or something like revocable tokens.

**Comment by Minoru** [comment]
> In Mutt you just have to put the statement in backticks instead of quotes. But IMHO this is not as straight forward, this might be familiar to programmers but could be confusing to some less technical users.

Whoa! Am a programmer, familiar enough with shell and Perl, been using Mutt for years, but didn't think of backticks :)

> I personally liked the term "eval" since this emphasizes on the evaluation/result of the given statement, not on the execution of some command.

Makes sense. I also looked through Newsbeuter settings and found that the only one that returns something into Newsbeuter is `html-renderer`; this means we don't have a precedent, and can do whatever we want :) Let's go with the "eval" then.

**Comment by x4121** [comment]
> Whoa! Am a programmer, familiar enough with shell and Perl, been using Mutt for years, but didn't think of backticks :)

Exactly, that's why I dislike this idea ;)
Also this suggests the usage of this feature in other places that might not be intended/covered and could result in hard to find bugs.

**Comment by x4121** [comment]
Oh wow... back to the "backticks": while I was generating/reading the documentation, I found (3 paragraphs before the "Table 1. Configuration Commands") that this should already be implemented

> It’s even possible to integrate the output of external commands into the configuration. The text between two backticks ("`") is evaluated as shell command, and its output is put on its place instead.

How could I've missed this part :/
Otherwise I can't seem to get it working, for an easy example (just echoing the parameter) I get

```
Loading configuration...Error while processing command `confirm-exit `echo yes`' (/home/armin/.newsbeuter/config line 1): expected boolean value, found ``echo' instead
```
So maybe this was planned/started but never finished.

**Comment by Minoru** [comment]
> How could I've missed this part :/

I knew about it, but it didn't occur to me even when you mentioned Mutt uses backticks to do "eval" thing =\

It's also interesting to note that the backticks [were implemented](https://github.com/akrennmair/newsbeuter/commit/8f6916c05bebf46b25fcf06d83e8fdfb75f1997a) years before [The Old Reader support came around](https://github.com/akrennmair/newsbeuter/commit/939288339c65312ed277d4bbafca9f9598ada832). So apparently the overlap between features didn't come to Andreas' mind, either.

I filed a separate issue about backticks not working, see #504. Thanks!

**Comment by x4121** [comment]
Doesn't this make the PR kinda redundant? Instead of `passwordeval "command"` I could just use ```password "`command`"```. Even having `passwordfile` seems a bit strange to me if you could just use ```password "`cat $file`"```.

**Comment by Minoru** [comment]
I don't think it does. As we discussed earlier, backticks aren't as obvious to users as straightforward options are.

They also make a config look more like a program rather than a simple `key = value` list. We have just a few places (e.g. `ignore-article`) where more complexity is warranted; the rest of it I'd rather keep as simple as possible.

Backticks are already there, so I can't just take them out, but I also don't think it's a good idea to advertise them as a way to solve ordinary problems. So if you're okay with finishing the PR, I'll be happy to merge it.

**Comment by x4121** [comment]
OK, got you. I'll try to get this done till the end of the week :)

**Comment by Minoru** [comment]
Closed via #503.
